### ISSUE TITLE
Unavailable in old devices

### ISSUE THREAD

**Issue by awaterho** [molstar/molstar#542]
Hi there,

Sorry to leave a vague comment here, but I find it difficult to get javascript messages from old devices. 

We have just updated ModelArchive to use the latest Mol* code, as we knew it fixed a cif parsing error, and also the default quality colours were desired. That's great!

However, on the downside, the latest code is twice as large (4.26MB vs 2.31MB), which is not really a problem, but we find it extremely slow (minutes) to startup on old iPads and old Android tablets report that webgl is not enabled.

Sorry to not have anymore useful details here, and I understand you want to take the project forwards without having to continue to support old devices, but I think I am just letting you know that molstar is no longer running on old devices. whereas NGL / PV continue to be available. 

Andrew

**Comment by dsehnal** [comment]
Hi Andrew,

thank you for the feedback. 

While it might be convenient for many users, the ``Viewer`` wrapper isn't intended for every use case. It is just a single instance of using the Mol* plugin.

One way the size of the viewer could be reduced to 2.6MB (0.7MB gzip) is by not including optional extensions. This would require a bit of work on your part by removing unused extensions [``src/apps/viewer/app.tx``](https://github.com/molstar/molstar/blob/master/src/apps/viewer/app.ts#L58), their corresponding imports and creating a custom build with WebPack. The biggest extension is the ``mp4-export`` which adds I think around 1.3MB to the code. It is also possible to import this extension on demand if you wish with a bit more WebPack configuration using dynamic imports.

Having said that, the package size could definitely use an improvement/lazy loading/more modularity for the code regarding imports and if we find a nice way to implement that, we will do so.

David

**Comment by awaterho** [comment]
I was able to remove the mp4-export and turn on gzip compression for the server - bringing it down to 0.7MB as you say. Further size reduction is possible if we enable brotli compression. 

Unfortunately that did not reduce the slow startup time on our old ipad. Just a note, the old android device is able to intialize a molstar window quickly, but fails to render anything to the canvas.

For now, I will close this issue, until I can do some debugging and produce some more useful info for you.

**Comment by dsehnal** [comment]
Thanks for the info @awaterho.

**Comment by arose** [comment]
Hey @awaterho, you might want to give #552 a try. The slowness you experience could be because an old device with an updated browser might use a software renderer due to GPU driver problems, which the PR should show you with an error. One thing to try is to force use of webgl1 (because you said it works in NGL which does not support webgl2). You can do so with the `prefer-webgl1=1` GET param in the viewer or the corresponding PluginConfig option.

**Comment by awaterho** [comment]
Hi Alex, 
I had a little time to look further into this, but sadly the webgl1 trick made little difference.
I found that in webgl.extensions.ts the errors Could not find support for "instanced_arrays" and also Could not find support for "element_index_uint" would be thrown, and if ignoring this then I end up with a webgl error "Need MAX_VERTEX_TEXTURE_IMAGE_UNITS" >=8. So this is all on the old android device, chrome latest possible which is 81.0.4144.138. Android system 3.3.2

Trying the allowMajorPerformanceCaveat=true made little difference to the startup time for the old ipad, including the URL you mention above. It's strange, it takes 1min 12 sec to show the workspace - but after that, I can still load the zika virus in an acceptable time (a few seconds). 

So really, I don't expect any time to be spent trying to fix things for old devices.

**Comment by dsehnal** [comment]
@awaterho do you think you would be able to test if the issue is parsing of the JS code or the plugin initialization itself? 

This could be achieved by doing 

```html
<script type="text/javascript">
  var start = Date.now();
  var viewer = new molstar.Viewer();
  document.body.innerText = "Empty Init: " + (Date.now() - start) + " ms";
</script>
```

vs 

```html
<script type="text/javascript">
  var start = Date.now();
  var viewer = new molstar.Viewer.create('app').then(function () {
      document.body.innerText = "Viewer Init: " + (Date.now() - start) + " ms";
  });
</script>
```

The issue might not be 3D visualization related and perhaps this could help up isolate what is causing it.

**Comment by awaterho** [comment]
iPad test - the first empty init was 22ms - 25ms;  The second viewer init was 113ms - 223ms. 

(If I roll back to molstar we were previously using, it still works well. NGL / PV still are very fast on this device)

**Comment by dsehnal** [comment]
Thanks!
So the error is likely some static initialization code. What was the previous Mol* version you were using?

**Comment by dsehnal** [comment]
Going to re-open this for now.

@awaterho can I please ask one more favor? Can you run ``embedded.html`` from https://github.com/molstar/molstar/pull/563. This will get the timing for loading all modules and perhaps we can find where the issue is.

**Comment by arose** [comment]
> Could not find support for "instanced_arrays" and also Could not find support for "element_index_uint" would be thrown, and if ignoring this then I end up with a webgl error "Need MAX_VERTEX_TEXTURE_IMAGE_UNITS" >=8.

These are hard requirements for molstar.

**Comment by awaterho** [comment]
Hello David,

I’ve been away a few days, and I came back to this now, but I am sorry, I am not able to debug Molstar so easily. I set debugMode and timingMode to true, but I do not see the output that you described.

In fact, I see very different output in Chrome and Firefox, so I am sorry,  I do not know enough about debugging Molstar to be any help right now.

I would say again it is too much trouble for you to debug on old devices.
